modules/openthread/platform/diag.c

```diff
@@ -68,3 +74,156 @@ void otPlatDiagAlarmCallback(otInstance *aInstance)
 {
 	ARG_UNUSED(aInstance);
 }
+
+/*
+ * To enable gpio diag commands, in Devicetree create `gpio_ports` node in root path
+ * with `compatible = "gpio-ports"` property and `gpio-ports-phandles` property,
+ * which should contain phandles of GPIO port nodes. e.g:
+ *
+ * gpio-ports {
+ *		compatible = "gpio-ports";
+ *		gpio-ports-phandles = <&gpio0 &gpio1>;
+ * };
+ *
+ * To enable reading current gpio pin mode, define
+ * `CONFIG_GPIO_GET_DIRECTION` in prj.conf.
+ *
+ * Note: `<gpio>` in `diag gpio` commands is an integer that combines
+ * port and pin into a single, contiguous number space as follows:
+ * gpio = (port * 32) + pin
+ */
+#if DT_NODE_EXISTS(DT_PATH(gpio_ports))
+
+static const struct device *const gpio_port_dev[] = {DT_FOREACH_PROP_ELEM_SEP(
+	DT_PATH(gpio_ports), gpio_ports_phandles, DT_GPIO_PORT_FROM_PROP_IDX, (,))};
+
+static otError gpio_get_dev_and_pin(uint32_t aGpio, gpio_pin_t *pin, const struct device **dev)
+{
+	otError error = OT_ERROR_NONE;
+	uint32_t port;
+
+	port = aGpio / GPIO_MAX_PINS_PER_PORT;
+	*pin = aGpio % GPIO_MAX_PINS_PER_PORT;
+	otEXPECT_ACTION(port < otARRAY_LENGTH(gpio_port_dev), error = OT_ERROR_INVALID_ARGS);
+
+	*dev = gpio_port_dev[port];
+	otEXPECT_ACTION(otPlatDiagModeGet(), error = OT_ERROR_INVALID_STATE);
+	otEXPECT_ACTION(device_is_ready(*dev), error = OT_ERROR_INVALID_ARGS);
+
+	const struct gpio_driver_config *const cfg =
+		(const struct gpio_driver_config *)((*dev)->config);
+
+	otEXPECT_ACTION((cfg->port_pin_mask & (gpio_port_pins_t)BIT(*pin)) != 0U,
+			error = OT_ERROR_INVALID_ARGS);
+
+exit:
+	return error;
+}
+
+otError otPlatDiagGpioSet(uint32_t aGpio, bool aValue)
+{
+	const struct device *dev;
+	otError error = OT_ERROR_NONE;
+	gpio_pin_t pin;
+	int rv;
+
+	error = gpio_get_dev_and_pin(aGpio, &pin, &dev);
+
+	otEXPECT(error == OT_ERROR_NONE);
+
+#if defined(CONFIG_GPIO_GET_DIRECTION)
+	rv = gpio_pin_is_output(dev, pin);
+	otEXPECT_ACTION(rv == 1, error = OT_ERROR_FAILED);
+#endif
+
+	rv = gpio_pin_set(dev, pin, (int)aValue);
+	otEXPECT_ACTION(rv == 0, error = OT_ERROR_FAILED);
+
+exit:
+	return error;
+}
+
+otError otPlatDiagGpioGet(uint32_t aGpio, bool *aValue)
+{
+	const struct device *dev;
+	otError error = OT_ERROR_NONE;
+	gpio_pin_t pin;
+	int rv;
+
+	error = gpio_get_dev_and_pin(aGpio, &pin, &dev);
+
+	otEXPECT(error == OT_ERROR_NONE);
+	otEXPECT_ACTION(aValue != NULL, error = OT_ERROR_INVALID_ARGS);
+
+#if defined(CONFIG_GPIO_GET_DIRECTION)
+	rv = gpio_pin_is_input(dev, pin);
+	otEXPECT_ACTION(rv == 1, error = OT_ERROR_FAILED);
+#endif
+
+	rv = gpio_pin_get(dev, pin);
+	otEXPECT_ACTION(rv >= 0, error = OT_ERROR_FAILED);
+	*aValue = (bool)rv;
+
+exit:
+	return error;
+}
+
+otError otPlatDiagGpioSetMode(uint32_t aGpio, otGpioMode aMode)
```

Comment:
It looks like we just need a diagnostic GPIO, so I think something like this https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/options/u-boot.yaml would make more sense:

```
options {
    openthread {
        compatible = "openthread,config";
        diag-gpios = <&gpio0 1 ...>;
    };
};

Reply:
@sjg20 could you comment on that since you're the maintainer of https://github.com/devicetree-org/dt-schema/blob/main/dtschema/schemas/options.yaml ?

Reply:
The issue is that `/options` node has not been used in Zephyr before and doesn't a proper description yet. Do you think we could require `openthread` node to have a label, e.g:

``` 
openthread_config: openthread {
        compatible = "openthread,config";
        diag-gpios = <&gpio0 3  GPIO_ACTIVE_HIGH>,
                     <&gpio1 11 GPIO_ACTIVE_HIGH>;
};
```

then we could use `DT_NODELABEL(openthread_config)` in diag.c and leave node's placement up to user for now. Later we could discuss proper placement for this type of nodes as a separate task?

Reply:
Even if not defined, it can be created. Whether additional DT helpers for such node should exist needs to be decided.